src/ray/gcs/gcs_server/gcs_task_manager.cc

```diff
@@ -418,5 +492,22 @@ void GcsTaskManager::RecordMetrics() {
       task_event_storage_->GetTaskEventsBytes());
 }
 
+void GcsTaskManager::OnJobFinished(const JobID &job_id, int64_t job_finish_time_ms) {
+  RAY_LOG(DEBUG) << "Marking all running tasks of job " << job_id.Hex() << " as failed.";
+  timer_.expires_from_now(boost::posix_time::milliseconds(
+      RayConfig::instance().gcs_mark_task_failed_on_job_done_delay_ms()));
```

Comment:
Is there any way to know if the job has failed / succeeded? I think we the job has succeeded, it is not necessary to mark children tasks as failed? (they should be eventually reported?)

Reply:
if a job succeeds and finishes, any running tasks will be killed, any finished tasks should have their core workers shutdown properly (which will report any task events before exit).

Shouldn't we marked any running tasks failed since they will be killed? E.g. a `time.sleep` that's never awaited and the job finishes.

Reply:
Do you happen to know where's the logic where raylet kills workers upon the driver exits? I thought this should do the proper cleanup (graceful). 

But it makes sense to do for both cases

Reply:
Looks like a worker gets killed if the owning worker dies here: 
https://github.com/ray-project/ray/blob/302a7e5671b51530c08944a505b145af81cdebc7/src/ray/raylet/node_manager.cc#L1074-L1077